File: src/com/app/test/viewpager/BrushCourseInfoCircle.java
package com.app.test.viewpager;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.app.test.R;
import com.app.test.util.DisplayUtil;

/**
 * Created by dell on 2017/10/12.
 * 圆圈进度条
 */

public class BrushCourseInfoCircle extends View {
    private Paint mPaint;
    private int mKeduLength;
    private int mProgressRingThick;
    private Bitmap mIndicatorBmp;
    private int mKedu2ProgressRingDistance;
    private int mEdge2KeduDistance;
    private int mPointDiscreteDistance;
    private int mWidth;
    private int mHeight;
    private float rate;
    private int KeduCount;
    private int mPointRadio;
    private float mMaxRadious;
    private OnSeekChangeListener mOnSeekArcChangeListener;
    private float thumbX;
    private float thumbY;
    private float mTouchIgnoreRadius;
    private boolean ignoreContinueMoveTouch;
    private float currentAngle = -1;
    private Matrix mThumbnailMatrix;
    private boolean isTracking; //用户是否在拖动
    private Path mTrackingPath;
    private Paint mTrackingPaint;
    private float beginTrackingAngle;
    private int mMirrorRadious;
    private float mMirrorRate;
    private Path mMirrorPath;
    private int mMPoint2EdgeDistance;

    public BrushCourseInfoCircle(Context context) {
        super(context);
        init();
    }

    public BrushCourseInfoCircle(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public BrushCourseInfoCircle(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setDither(true);

        //虚线
        mTrackingPaint = new Paint(mPaint);
        mTrackingPaint.setStyle(Paint.Style.STROKE);
        mTrackingPaint.setColor(Color.YELLOW);
        mTrackingPaint.setStrokeWidth(DisplayUtil.getInstance().dip2px(1));
        mTrackingPaint.setPathEffect(new DashPathEffect(new float[]{DisplayUtil.getInstance().dip2px(2), DisplayUtil.getInstance().dip2px(2)}, 0));
        mTrackingPath = new Path();

        //放大镜
        mMirrorRadious = DisplayUtil.getInstance().dip2px(20);        //放大镜半斤
        mMirrorRate = 1.5f;        //放大镜放大的倍速
        mMirrorPath = new Path();

        //整体缩放比率
        rate = 1;
        //指示器触摸偏差
        mTouchIgnoreRadius = DisplayUtil.getInstance().dip2px(20);
        //刻度数
        KeduCount = 36;
        //刻度线长度
        mKeduLength = DisplayUtil.getInstance().dip2px(5);
        //进度环厚度
        mProgressRingThick = DisplayUtil.getInstance().dip2px(15);

        //视频点的半径
        mPointRadio = DisplayUtil.getInstance().dip2px(5);
        //视频点扩散的距离
//        mPointDiscreteDistance = DisplayUtil.dip2px(getContext(),10);
        mPointDiscreteDistance = 0;
        //视频点圆形 到 最外灰圈的距离
        mMPoint2EdgeDistance = DisplayUtil.getInstance().dip2px(7);
        //最外圈灰线到刻度内圆边距的距离
        mEdge2KeduDistance = DisplayUtil.getInstance().dip2px(5) + mKeduLength;
        //刻度线内圆边距到进度环外圆边距的距离
        mKedu2ProgressRingDistance = mProgressRingThick / 2 + DisplayUtil.getInstance().dip2px(3);
        //进度指示器
        mIndicatorBmp = BitmapFactory.decodeResource(getResources(), R.drawable.cg_indicator);
        mThumbnailMatrix = new Matrix();

    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        mWidth = getMeasuredWidth();
        mHeight = getMeasuredHeight();
        mMaxRadious = Math.min(mWidth, mHeight) / 2 * rate - DisplayUtil.getInstance().dip2px(20);//最大半径 (20:指示器超出的长度 )

        canvas.translate(mWidth / 2, mHeight / 2);//平移圆心
        drawBg(canvas);
        drawKeDu(canvas);
        drawPoint(canvas);

        //放大镜
        if (isTracking) {
            canvas.save();
            //放大镜圆形离大圆盘圆心的距离
            float mirrorDiscreteRadio = mMaxRadious - 2.0f * (float) mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - (float) mProgressRingThick / 2.0f;
            mMirrorPath.reset();
            mMirrorPath.addCircle(getRealCosX(currentAngle, mirrorDiscreteRadio), getRealSinY(currentAngle, mirrorDiscreteRadio), mMirrorRadious, Path.Direction.CW);
            canvas.clipPath(mMirrorPath);

            drawBg(canvas);//画没放大的背景
            drawKeDu(canvas);//画没放大的刻度
            //画放大的进度
            float scaleOneRadious = mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - (float) mProgressRingThick / 2.0f;
            float X = getRealCosX(currentAngle, scaleOneRadious * (mMirrorRate - 1.0f));
            float Y = getRealSinY(currentAngle, scaleOneRadious * (mMirrorRate - 1.0f));
            canvas.translate(-X, -Y);
            canvas.scale(mMirrorRate, mMirrorRate);

            canvas.restore();
        }

        drawThumbnail(canvas);

    }


    /*画指示器*/
    private void drawThumbnail(Canvas canvas) {
        if (currentAngle >= 0) {
            //画瞄准的虚线
            if (isTracking) {
                canvas.save();
                canvas.rotate(currentAngle);
                mTrackingPath.reset();
                mTrackingPath.moveTo(0, -mMaxRadious);
                mTrackingPath.lineTo(0, -mMaxRadious + (mPointDiscreteDistance + mEdge2KeduDistance + mEdge2KeduDistance + mKedu2ProgressRingDistance + mProgressRingThick));
                canvas.drawPath(mTrackingPath, mTrackingPaint);
                canvas.restore();
            }

            thumbX = getRealCosX(currentAngle, mMaxRadious);
            thumbY = getRealSinY(currentAngle, mMaxRadious);

            if (mIndicatorBmp != null) {
                int bmpWidth = mIndicatorBmp.getWidth();
                int bmpHeight = mIndicatorBmp.getHeight();

                canvas.save();
                canvas.translate(thumbX - (float) bmpWidth / 2.0f, thumbY - (float) bmpHeight / 2.0f);
                mThumbnailMatrix.setRotate(currentAngle, (float) bmpWidth / 2.0f, (float) bmpHeight / 2.0f);
                canvas.drawBitmap(mIndicatorBmp, mThumbnailMatrix, null);
                canvas.restore();
            }
        }
    }

    private void drawKeDu(Canvas canvas) {
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeWidth(DisplayUtil.getInstance().dip2px(1));

        canvas.save();
        float startY = -(mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance);
        float stopY = startY - mKeduLength;
        float angle = (float) 360 / KeduCount;
        float accuAngle = 0;
        for (int i = 0; i <= KeduCount; i++) {
            if (accuAngle <= currentAngle) {
                mPaint.setColor(Color.parseColor("#ffffff"));
            } else {
                mPaint.setColor(Color.parseColor("#1affffff"));
            }
            canvas.drawLine(0, startY, 0, stopY, mPaint);
            canvas.rotate(angle);
            accuAngle += angle;
        }
        canvas.restore();
    }

    private void drawPoint(Canvas canvas) {
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(Color.parseColor("#22CA92"));
        float angle = 0;
        if (angle >= 0 && angle <= 360) {
            canvas.save();
            canvas.rotate(angle);
            canvas.drawCircle(0, -(mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mMPoint2EdgeDistance - mPointRadio), mPointRadio, mPaint);
            canvas.restore();
        }

    }


    /*画静态背景*/
    private void drawBg(Canvas canvas) {
        //画背景
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(Color.parseColor("#21ffffff"));
        canvas.drawCircle(0, 0, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance, mPaint);

        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setColor(Color.parseColor("#1affffff"));
        mPaint.setStrokeWidth(DisplayUtil.getInstance().dip2px(1));

        //话最外面的灰圈
        canvas.drawCircle(0, 0, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance, mPaint);
        //画最里边的圈
        canvas.drawCircle(0, 0, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - mProgressRingThick, mPaint);

        //画进度环
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setColor(Color.parseColor("#333333"));
        mPaint.setStrokeWidth(mProgressRingThick);
        canvas.drawCircle(0, 0, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - (float) mProgressRingThick / 2.0f, mPaint);

    }


    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (isEnabled()) {
            this.getParent().requestDisallowInterceptTouchEvent(true);
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    updateOnTouch(event);
                    break;
                case MotionEvent.ACTION_MOVE:
                    updateOnTouch(event);
                    break;
                case MotionEvent.ACTION_UP:
                    if (!ignoreContinueMoveTouch) {
                        onStopTrackingTouch();
                    }
                    ignoreContinueMoveTouch = true;
                    setPressed(false);
                    this.getParent().requestDisallowInterceptTouchEvent(false);
                    break;
                case MotionEvent.ACTION_CANCEL:
                    if (!ignoreContinueMoveTouch) {
                        onStopTrackingTouch();
                    }
                    ignoreContinueMoveTouch = true;
                    setPressed(false);
                    this.getParent().requestDisallowInterceptTouchEvent(false);
                    break;
            }
            return true;
        }
        return false;
    }

    private void updateOnTouch(MotionEvent event) {
        boolean ignoreTouch = ignoreTouch(event.getX(), event.getY());
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            ignoreContinueMoveTouch = ignoreTouch;
            if (ignoreTouch) {
                return;
            }
            onStartTrackingTouch();
        }
        if (event.getAction() == MotionEvent.ACTION_MOVE) {
            //down事件忽略,后面的move事件也忽略
            if (ignoreContinueMoveTouch) {
                return;
            }
        }
        setPressed(true);
        float touchAngle = getTouchDegrees(event.getX(), event.getY());
        onProgressRefresh(touchAngle, true);
    }

    public void setCurrentAngle(float angle) {
        if (isTracking) {
            return;
        }
        this.currentAngle = angle;
    }

    public float getCurrentAngle() {
        return currentAngle;
    }

    /*还原到上一次track的位置*/
    public void revertTrack() {
        setCurrentAngle(beginTrackingAngle);
        invalidate();
    }

    private void onProgressRefresh(float angle, boolean isFromUser) {
        if (angle < 0 || angle > 360) {
            return;
        }
        this.currentAngle = angle;
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener.onProgressChanged(this, angle, isFromUser);
        }
        invalidate();
    }


    private float getTouchDegrees(float xPos, float yPos) {
        float x = xPos - mWidth / 2;
        float y = yPos - mHeight / 2;
        float angle = (float) Math.toDegrees(Math.atan2(y, x) + (Math.PI / 2));
        if (angle < 0) {
            angle = 360 + angle;
        }
        return angle;
    }

    private boolean ignoreTouch(float xPos, float yPos) {
        boolean ignore = false;
        float diffX = xPos - mWidth / 2 - thumbX;
        float diffY = yPos - mHeight / 2 - thumbY;
        float touchRadius = (float) Math.sqrt(((diffY * diffY) + (diffX * diffX)));
//        System.out.println("触摸距离========" + DisplayUtil.px2dip(getContext(),touchRadius));
        if (touchRadius > mTouchIgnoreRadius) {
            ignore = true;
        }
        return ignore;
    }

    private void onStopTrackingTouch() {
        this.isTracking = false;
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener.onStopTrackingTouch(this);
        }
        invalidate();
    }

    private void onStartTrackingTouch() {
        this.isTracking = true;
        this.beginTrackingAngle = currentAngle;
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener.onStartTrackingTouch(this);
        }
    }

    private float getRealCosX(float angle, float radius) {
        if (angle < 0 || angle > 360) {
            return 0;
        }
        if (angle < 90) { // 一
            return getCosX(90 - angle, radius);
        } else if (angle < 180) { //四
            return getCosX(angle - 90, radius);
        } else if (angle < 270) { //三
            return -getCosX(270 - angle, radius);
        } else { //二
            return -getCosX(angle - 270, radius);
        }
    }

    private float getRealSinY(float angle, float radius) {
        if (angle < 0 || angle > 360) {
            return 0;
        }
        if (angle < 90) { // 一
            return -getSinY(90 - angle, radius);
        } else if (angle < 180) { //四
            return getSinY(angle - 90, radius);
        } else if (angle < 270) { //三
            return getSinY(270 - angle, radius);
        } else { //二
            return -getSinY(angle - 270, radius);
        }
    }


    private float getCosX(float diggre, float radius) { //diggre  0-90 第一象限角度计算
        return (float) (radius * Math.cos(Math.toRadians(diggre)));
    }

    private float getSinY(float diggre, float radius) { //diggre  0-90
        return (float) (radius * Math.sin(Math.toRadians(diggre)));
    }


    public void setOnSeekArcChangeListener(OnSeekChangeListener l) {
        mOnSeekArcChangeListener = l;
    }

    public void setPointList() {
    }

    /*设置作业点的位置*/
    public void setWorkPointList() {

    }

    public void updateView() {
        invalidate();
    }

    public interface OnSeekChangeListener {

        void onProgressChanged(BrushCourseInfoCircle seekCircle, float angle, boolean fromUser);


        void onStartTrackingTouch(BrushCourseInfoCircle seekCircle);


        void onStopTrackingTouch(BrushCourseInfoCircle seekCircle);
    }

}

File: src/com/app/test/viewpager/BrushCourseInfoCircle.kt
package com.app.test.viewpager

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.app.test.R
import com.app.test.util.DensityUtil.dp2px
import kotlin.math.cos
import kotlin.math.sin

/**
 * Created by dell on 2017/10/12.
 * 圆圈进度条
 */
class BrushCourseInfoCircle : View {
    private var mPaint: Paint? = null
    private var mKeduLength = 0
    private var mProgressRingThick = 0
    private var mIndicatorBmp: Bitmap? = null
    private var mKedu2ProgressRingDistance = 0
    private var mEdge2KeduDistance = 0
    private var mPointDiscreteDistance = 0
    private var mWidth = 0
    private var mHeight = 0
    private var rate = 0f
    private var KeduCount = 0
    private var mPointRadio = 0
    private var mMaxRadious = 0f
    private var mOnSeekArcChangeListener: OnSeekChangeListener? = null
    private var thumbX = 0f
    private var thumbY = 0f
    private var mTouchIgnoreRadius = 0f
    private var ignoreContinueMoveTouch = false
    private var currentAngle = -1f
    private var mThumbnailMatrix: Matrix? = null
    private var isTracking: Boolean = false//用户是否在拖动 = false
    private var mTrackingPath: Path? = null
    private var mTrackingPaint: Paint? = null
    private var beginTrackingAngle = 0f
    private var mMirrorRadious = 0
    private var mMirrorRate = 0f
    private var mMirrorPath: Path? = null
    private var mMPoint2EdgeDistance = 0

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mPaint!!.isDither = true

        //虚线
        mTrackingPaint = Paint(mPaint)
        mTrackingPaint!!.style = Paint.Style.STROKE
        mTrackingPaint!!.color = Color.YELLOW
        mTrackingPaint!!.strokeWidth = dp2px(1).toFloat()
        mTrackingPaint!!.pathEffect = DashPathEffect(floatArrayOf(dp2px(2).toFloat(), dp2px(2).toFloat()), 0f)
        mTrackingPath = Path()

        //放大镜
        mMirrorRadious = dp2px(20) //放大镜半斤
        mMirrorRate = 1.5f //放大镜放大的倍速
        mMirrorPath = Path()

        //整体缩放比率
        rate = 1f
        //指示器触摸偏差
        mTouchIgnoreRadius = dp2px(20).toFloat()
        //刻度数
        KeduCount = 36
        //刻度线长度
        mKeduLength = dp2px(5)
        //进度环厚度
        mProgressRingThick = dp2px(15)

        //视频点的半径
        mPointRadio = dp2px(5)
        mPointDiscreteDistance = 0
        //视频点圆形 到 最外灰圈的距离
        mMPoint2EdgeDistance = dp2px(7)
        //最外圈灰线到刻度内圆边距的距离
        mEdge2KeduDistance = dp2px(5) + mKeduLength
        //刻度线内圆边距到进度环外圆边距的距离
        mKedu2ProgressRingDistance = mProgressRingThick / 2 + dp2px(3)
        //进度指示器
        mIndicatorBmp = BitmapFactory.decodeResource(resources, R.drawable.cg_indicator)
        mThumbnailMatrix = Matrix()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        mWidth = measuredWidth
        mHeight = measuredHeight
        mMaxRadious = Math.min(mWidth, mHeight) / 2 * rate - dp2px(20) //最大半径 (20:指示器超出的长度 )
        canvas.translate(mWidth / 2.toFloat(), mHeight / 2.toFloat()) //平移圆心
        drawBg(canvas)
        drawKeDu(canvas)
        drawPoint(canvas)

        //放大镜
        if (isTracking) {
            canvas.save()
            //放大镜圆形离大圆盘圆心的距离
            val mirrorDiscreteRadio = mMaxRadious - 2.0f * mPointRadio.toFloat() - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - mProgressRingThick.toFloat() / 2.0f
            mMirrorPath!!.reset()
            mMirrorPath!!.addCircle(getRealCosX(currentAngle, mirrorDiscreteRadio), getRealSinY(currentAngle, mirrorDiscreteRadio), mMirrorRadious.toFloat(), Path.Direction.CW)
            canvas.clipPath(mMirrorPath)
            drawBg(canvas) //画没放大的背景
            drawKeDu(canvas) //画没放大的刻度
            //画放大的进度
            val scaleOneRadious = mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - mProgressRingThick.toFloat() / 2.0f
            val X = getRealCosX(currentAngle, scaleOneRadious * (mMirrorRate - 1.0f))
            val Y = getRealSinY(currentAngle, scaleOneRadious * (mMirrorRate - 1.0f))
            canvas.translate(-X, -Y)
            canvas.scale(mMirrorRate, mMirrorRate)
            canvas.restore()
        }
        drawThumbnail(canvas)
    }

    /*画指示器*/
    private fun drawThumbnail(canvas: Canvas) {
        if (currentAngle >= 0) {
            //画瞄准的虚线
            if (isTracking) {
                canvas.save()
                canvas.rotate(currentAngle)
                mTrackingPath!!.reset()
                mTrackingPath!!.moveTo(0f, -mMaxRadious)
                mTrackingPath!!.lineTo(0f, -mMaxRadious + (mPointDiscreteDistance + mEdge2KeduDistance + mEdge2KeduDistance + mKedu2ProgressRingDistance + mProgressRingThick))
                canvas.drawPath(mTrackingPath, mTrackingPaint)
                canvas.restore()
            }
            thumbX = getRealCosX(currentAngle, mMaxRadious)
            thumbY = getRealSinY(currentAngle, mMaxRadious)
            if (mIndicatorBmp != null) {
                val bmpWidth = mIndicatorBmp!!.width
                val bmpHeight = mIndicatorBmp!!.height
                canvas.save()
                canvas.translate(thumbX - bmpWidth.toFloat() / 2.0f, thumbY - bmpHeight.toFloat() / 2.0f)
                mThumbnailMatrix!!.setRotate(currentAngle, bmpWidth.toFloat() / 2.0f, bmpHeight.toFloat() / 2.0f)
                canvas.drawBitmap(mIndicatorBmp, mThumbnailMatrix, null)
                canvas.restore()
            }
        }
    }

    private fun drawKeDu(canvas: Canvas) {
        mPaint!!.style = Paint.Style.STROKE
        mPaint!!.strokeWidth = dp2px(1).toFloat()
        canvas.save()
        val startY = -(mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance)
        val stopY = startY - mKeduLength
        val angle = 360.toFloat() / KeduCount
        var accuAngle = 0f
        for (i in 0..KeduCount) {
            if (accuAngle <= currentAngle) {
                mPaint!!.color = Color.parseColor("#ffffff")
            } else {
                mPaint!!.color = Color.parseColor("#1affffff")
            }
            canvas.drawLine(0f, startY, 0f, stopY, mPaint)
            canvas.rotate(angle)
            accuAngle += angle
        }
        canvas.restore()
    }

    private fun drawPoint(canvas: Canvas) {
        mPaint!!.style = Paint.Style.FILL
        mPaint!!.color = Color.parseColor("#22CA92")
        val angle = 0f
        if (angle >= 0 && angle <= 360) {
            canvas.save()
            canvas.rotate(angle)
            canvas.drawCircle(0f, -(mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mMPoint2EdgeDistance - mPointRadio), mPointRadio.toFloat(), mPaint)
            canvas.restore()
        }
    }

    /*画静态背景*/
    private fun drawBg(canvas: Canvas) {
        //画背景
        mPaint!!.style = Paint.Style.FILL
        mPaint!!.color = Color.parseColor("#21ffffff")
        canvas.drawCircle(0f, 0f, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance, mPaint)
        mPaint!!.style = Paint.Style.STROKE
        mPaint!!.color = Color.parseColor("#1affffff")
        mPaint!!.strokeWidth = dp2px(1).toFloat()

        //话最外面的灰圈
        canvas.drawCircle(0f, 0f, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance, mPaint)
        //画最里边的圈
        canvas.drawCircle(0f, 0f, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - mProgressRingThick, mPaint)

        //画进度环
        mPaint!!.style = Paint.Style.STROKE
        mPaint!!.color = Color.parseColor("#333333")
        mPaint!!.strokeWidth = mProgressRingThick.toFloat()
        canvas.drawCircle(0f, 0f, mMaxRadious - 2 * mPointRadio - mPointDiscreteDistance - mEdge2KeduDistance - mKedu2ProgressRingDistance - mProgressRingThick.toFloat() / 2.0f, mPaint)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (isEnabled) {
            this.parent.requestDisallowInterceptTouchEvent(true)
            when (event.action) {
                MotionEvent.ACTION_DOWN -> updateOnTouch(event)
                MotionEvent.ACTION_MOVE -> updateOnTouch(event)
                MotionEvent.ACTION_UP -> {
                    if (!ignoreContinueMoveTouch) {
                        onStopTrackingTouch()
                    }
                    ignoreContinueMoveTouch = true
                    isPressed = false
                    this.parent.requestDisallowInterceptTouchEvent(false)
                }
                MotionEvent.ACTION_CANCEL -> {
                    if (!ignoreContinueMoveTouch) {
                        onStopTrackingTouch()
                    }
                    ignoreContinueMoveTouch = true
                    isPressed = false
                    this.parent.requestDisallowInterceptTouchEvent(false)
                }
            }
            return true
        }
        return false
    }

    private fun updateOnTouch(event: MotionEvent) {
        val ignoreTouch = ignoreTouch(event.x, event.y)
        if (event.action == MotionEvent.ACTION_DOWN) {
            ignoreContinueMoveTouch = ignoreTouch
            if (ignoreTouch) {
                return
            }
            onStartTrackingTouch()
        }
        if (event.action == MotionEvent.ACTION_MOVE) {
            //down事件忽略,后面的move事件也忽略
            if (ignoreContinueMoveTouch) {
                return
            }
        }
        isPressed = true
        val touchAngle = getTouchDegrees(event.x, event.y)
        onProgressRefresh(touchAngle, true)
    }

    fun setCurrentAngle(angle: Float) {
        if (isTracking) {
            return
        }
        currentAngle = angle
    }

    fun getCurrentAngle(): Float {
        return currentAngle
    }

    /*还原到上一次track的位置*/
    fun revertTrack() {
        setCurrentAngle(beginTrackingAngle)
        invalidate()
    }

    private fun onProgressRefresh(angle: Float, isFromUser: Boolean) {
        if (angle < 0 || angle > 360) {
            return
        }
        currentAngle = angle
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener!!.onProgressChanged(this, angle, isFromUser)
        }
        invalidate()
    }

    private fun getTouchDegrees(xPos: Float, yPos: Float): Float {
        val x = xPos - mWidth / 2
        val y = yPos - mHeight / 2
        var angle = Math.toDegrees(Math.atan2(y.toDouble(), x.toDouble()) + Math.PI / 2).toFloat()
        if (angle < 0) {
            angle = 360 + angle
        }
        return angle
    }

    private fun ignoreTouch(xPos: Float, yPos: Float): Boolean {
        var ignore = false
        val diffX = xPos - mWidth / 2 - thumbX
        val diffY = yPos - mHeight / 2 - thumbY
        val touchRadius = Math.sqrt((diffY * diffY + diffX * diffX).toDouble()).toFloat()
        //        System.out.println("触摸距离========" + DisplayUtil.px2dip(getContext(),touchRadius));
        if (touchRadius > mTouchIgnoreRadius) {
            ignore = true
        }
        return ignore
    }

    private fun onStopTrackingTouch() {
        isTracking = false
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener!!.onStopTrackingTouch(this)
        }
        invalidate()
    }

    private fun onStartTrackingTouch() {
        isTracking = true
        beginTrackingAngle = currentAngle
        if (mOnSeekArcChangeListener != null) {
            mOnSeekArcChangeListener!!.onStartTrackingTouch(this)
        }
    }

    private fun getRealCosX(angle: Float, radius: Float): Float {
        if (angle < 0 || angle > 360) {
            return 0f
        }
        return when {
            angle < 90 -> { // 一
                getCosX(90 - angle, radius)
            }
            angle < 180 -> { //四
                getCosX(angle - 90, radius)
            }
            angle < 270 -> { //三
                -getCosX(270 - angle, radius)
            }
            else -> { //二
                -getCosX(angle - 270, radius)
            }
        }
    }

    private fun getRealSinY(angle: Float, radius: Float): Float {
        if (angle < 0 || angle > 360) {
            return 0f
        }
        return when {
            angle < 90 -> { // 一
                -getSinY(90 - angle, radius)
            }
            angle < 180 -> { //四
                getSinY(angle - 90, radius)
            }
            angle < 270 -> { //三
                getSinY(270 - angle, radius)
            }
            else -> { //二
                -getSinY(angle - 270, radius)
            }
        }
    }

    private fun getCosX(diggre: Float, radius: Float): Float { //diggre  0-90 第一象限角度计算
        return (radius * cos(Math.toRadians(diggre.toDouble()))).toFloat()
    }

    private fun getSinY(diggre: Float, radius: Float): Float { //diggre  0-90
        return (radius * sin(Math.toRadians(diggre.toDouble()))).toFloat()
    }

    fun setOnSeekArcChangeListener(l: OnSeekChangeListener?) {
        mOnSeekArcChangeListener = l
    }

    fun setPointList() {}

    /*设置作业点的位置*/
    fun setWorkPointList() {}
    fun updateView() {
        invalidate()
    }

    interface OnSeekChangeListener {
        fun onProgressChanged(seekCircle: BrushCourseInfoCircle?, angle: Float, fromUser: Boolean)
        fun onStartTrackingTouch(seekCircle: BrushCourseInfoCircle?)
        fun onStopTrackingTouch(seekCircle: BrushCourseInfoCircle?)
    }
}